File: macos/Services/MCP/Security/MCPIdentifierValidator.swift
// MCPIdentifierValidator.swift
// Gridex
//
// Allowlist check for SQL identifiers received from MCP clients. Defense-
// in-depth: even when the identifier is later quoted via
// `SQLDialect.quoteIdentifier`, this validator rejects anything outside a
// strict ASCII shape so unusual names never reach a driver.

import Foundation

enum MCPIdentifierValidator {
    static let maxLength = 128

    static func isValid(_ identifier: String) -> Bool {
        var count = 0
        for scalar in identifier.unicodeScalars {
            count += 1
            if count > maxLength { return false }
            let v = scalar.value
            let isLetter = (v >= 0x41 && v <= 0x5A) || (v >= 0x61 && v <= 0x7A) || v == 0x5F
            let isDigit = v >= 0x30 && v <= 0x39
            if count == 1 {
                if !isLetter { return false }
            } else {
                if !isLetter && !isDigit { return false }
            }
        }
        return count > 0
    }

    static func validate(_ identifier: String, as name: String = "identifier") throws {
        guard isValid(identifier) else {
            throw MCPToolError.invalidParameters(
                "\(name) '\(identifier)' contains invalid characters. Allowed: letters, digits, underscore; must start with a letter or underscore; max \(maxLength) chars."
            )
        }
    }

    /// Shared extraction for MCP write tools: reads `table_name` and optional
    /// `schema` from params and validates both. The returned schema is `nil`
    /// when the caller omitted it.
    static func extractTableAndSchema(from params: JSONValue) throws -> (table: String, schema: String?) {
        guard let table = params["table_name"]?.stringValue else {
            throw MCPToolError.invalidParameters("table_name is required")
        }
        try validate(table, as: "table_name")

        let schema = params["schema"]?.stringValue
        if let schema { try validate(schema, as: "schema") }
        return (table, schema)
    }
}

File: windows/Gridex/Services/MCP/Security/MCPIdentifierValidator.cpp
//
// MCPIdentifierValidator.cpp
//

#include "MCPIdentifierValidator.h"
#include "../../../Models/MCP/MCPToolError.h"

namespace DBModels { namespace MCPIdentifierValidator {

bool isValid(const std::string& identifier)
{
    if (identifier.empty() || identifier.size() > static_cast<size_t>(kMaxLength))
        return false;

    for (size_t i = 0; i < identifier.size(); ++i)
    {
        const unsigned char c = static_cast<unsigned char>(identifier[i]);
        const bool isLetter = (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_';
        const bool isDigit  = c >= '0' && c <= '9';
        if (i == 0)
        {
            if (!isLetter) return false;
        }
        else
        {
            if (!isLetter && !isDigit) return false;
        }
    }
    return true;
}

void validate(const std::string& identifier, const std::string& name)
{
    if (!isValid(identifier))
    {
        throw MCPToolError::invalidParameters(
            name + " '" + identifier + "' contains invalid characters. "
            "Allowed: letters, digits, underscore; must start with a letter or underscore; max "
            + std::to_string(kMaxLength) + " chars.");
    }
}

TableSchema extractTableAndSchema(const nlohmann::json& params)
{
    TableSchema ts;
    if (!params.contains("table_name") || !params["table_name"].is_string())
        throw MCPToolError::invalidParameters("table_name is required");

    ts.table = params["table_name"].get<std::string>();
    validate(ts.table, "table_name");

    if (params.contains("schema") && params["schema"].is_string())
    {
        auto s = params["schema"].get<std::string>();
        validate(s, "schema");
        ts.schema = s;
    }
    return ts;
}

}} // namespace
